### ISSUE TITLE
MIDI causing NESizer to crash.

### ISSUE THREAD

**Issue by jeff464** [Jaffe-/NESizer2#29]
Exactly as the title says. When I hit play on my MPC Live, the NESizer blanks out. I can play notes all day long just fine, but as soon as I hit play, it's toast. Also worth noting, I can duplicate this behavior if I have my Roland MC-505 connected via MIDI and press any of the track mute buttons, but play and notes seem to send to the NESizer just fine. Also, Ableton notes/play work fine as well. I can duplicate the behavior on both of my NESizer units.

**Comment by beau-seidon** [comment]
if you have a usb midi interface, hook up the midi output of your mpc live to the interface input and use midi-ox to capture the message it sends when you press play. that will help with diagnosis. 

try to capture any other controllers' midi messages that you have seen crash nesizer as well.

**Comment by jeff464** [comment]
So it would seem as though the MPC no longer crashes the NESizer, but I think this *might* be due to a recent firmware update from Akai. However, I noticed that powering the Roland MC-505 caused the NESizer to crash if MIDI was connected. Here's what MIDIOX returned:

![midiox](https://github.com/user-attachments/assets/bd74cb30-6deb-479c-8984-98b23b12514c)

**Comment by jeff464** [comment]
Here is output from the Roland MC-505 when I press a track button - it sends a sysex message, and this is ultimately what causes the NESizer to crash.

![midiox2](https://github.com/user-attachments/assets/7a5907f7-4510-4d13-be9f-bc675d124547)

**Comment by beau-seidon** [comment]
I'll see if I can figure out how it handles sysex and make it ignore sysex messages not specifically meant for NESizer.

**Comment by beau-seidon** [comment]
I have added a statement which checks for the proper SysexID and DeviceID in the sysex message handler, and it seems to functioning correctly. 

However, I had to also edit the program which converts samples and sequences to .syx files, and I'm still testing that to make sure it's working properly and samples get uploaded to the right slots and play back etc. 

Will upload to a new branch on my fork for public testing when I get it working.

**Comment by beau-seidon** [comment]
https://github.com/beau-seidon/NESizer2/tree/sysex-filter

Working fine now on my end. 
I included the recompiled gensysex.exe file for converting samples to sysex with the ID's in the message header. 
Let me know if you have issues with sample loading, playback, or more hangs / midi-handling misbehavior.

One thing I noticed, is that on infrequent occasion, it seems to ignore a correctly addressed sysex message from MIDIOX, and I have to send it again. Doesn't happen often, and when I resend, it usually works the second time.

**Comment by jeff464** [comment]
Great - just uploaded, and the crashes seem to be fixed! That said, is there a trick to getting the MIDI CCs to work? It seems I'm not able to adjust anything via MIDI.

**Comment by jeff464** [comment]
Wait, sorry - just realizing these are two different forks. So I assume it's now a choice between crashes or MIDI CC?

**Comment by beau-seidon** [comment]
Ruh roh. Give me a little bit, I'm busy at the moment, but I will make the sysex changes to a clone of circuitcreature's fork and upload so you can play with it. It should be a pretty straightforward copy/paste process now. 

As far as the MIDI CC features, I haven't tested their fork very much on my own synth yet so I don't know what issues to expect. Let me know how it goes!

**Comment by beau-seidon** [comment]
> Wait, sorry - just realizing these are two different forks. So I assume it's now a choice between crashes or MIDI CC?

https://github.com/beau-seidon/NESizer2/tree/circuitcreature-fix-sysex

**Comment by beau-seidon** [comment]
Adding the ignore function based on SysEx messages' "ID bytes" fixed the crashes from irrelevant sysex messages, but I found that sometimes the synth is missing bytes (race condition?) which can result in switching to the IGNORE state erroneously.

I'm working on a branch that has already fixed that issue but I still have a couple more things to test. Might have it committed this evening.

I also added patch selection via SysEx which works just like ProgramChange.
